File: benchmarks/heavy_benchmark.go
// Go HEAVY BENCHMARK
package main

import (
	"fmt"
	"time"
)

// ============================================================================
// TEST 1: FIBONACCI
// ============================================================================

func fib(n int64) int64 {
	if n <= 1 {
		return n
	}
	return fib(n-1) + fib(n-2)
}

func benchFib() int64 {
	start := time.Now()
	result := fib(40)
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("Fibonacci(40) = %d in %dms\n", result, elapsed)
	return elapsed
}

// ============================================================================
// TEST 2: MEMORY ALLOCATION
// ============================================================================

func benchAlloc(count int) int64 {
	start := time.Now()
	slices := make([][]byte, count)
	for i := 0; i < count; i++ {
		slices[i] = make([]byte, 1024)
		slices[i][0] = byte(i % 256)
	}
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("Allocated %d x 1KB blocks in %dms\n", count, elapsed)
	return elapsed
}

// ============================================================================
// TEST 3: CONCURRENT SIMULATION
// ============================================================================

func processRequest(id int) int64 {
	var sum int64 = 0
	for i := 0; i < 10000; i++ {
		sum += int64(i)
	}
	return sum
}

func benchConcurrent(requests int) int64 {
	start := time.Now()
	completed := 0

	for i := 0; i < requests; i++ {
		result := processRequest(i)
		if result > 0 {
			completed++
		}
	}

	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("Processed %d requests in %dms\n", completed, elapsed)
	return elapsed
}

// ============================================================================
// TEST 4: PRIME SIEVE
// ============================================================================

func sieve(limit int) int {
	isPrime := make([]bool, limit)
	for i := range isPrime {
		isPrime[i] = true
	}
	isPrime[0] = false
	isPrime[1] = false

	for i := 2; i*i < limit; i++ {
		if isPrime[i] {
			for j := i * i; j < limit; j += i {
				isPrime[j] = false
			}
		}
	}

	count := 0
	for _, p := range isPrime {
		if p {
			count++
		}
	}
	return count
}

func benchSieve() int64 {
	start := time.Now()
	primes := sieve(10000000)
	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("Found %d primes under 10M in %dms\n", primes, elapsed)
	return elapsed
}

// ============================================================================
// TEST 5: VECTOR OPERATIONS
// ============================================================================

func benchVector(size int) int64 {
	start := time.Now()

	a := make([]int64, size)
	b := make([]int64, size)
	c := make([]int64, size)

	for i := 0; i < size; i++ {
		a[i] = int64(i)
		b[i] = int64(i * 2)
	}

	for i := 0; i < size; i++ {
		c[i] = a[i] + b[i]
	}

	elapsed := time.Since(start).Milliseconds()
	fmt.Printf("Vector add of %d elements in %dms\n", size, elapsed)
	_ = c // prevent optimization
	return elapsed
}

// ============================================================================
// MAIN
// ============================================================================

func main() {
	fmt.Println("=== GO HEAVY BENCHMARK ===\n")

	fmt.Println("TEST 1: Fibonacci (Recursive)")
	benchFib()

	fmt.Println("\nTEST 2: Memory Allocation (1M blocks)")
	benchAlloc(1000000)

	fmt.Println("\nTEST 3: Concurrent Requests (100K)")
	benchConcurrent(100000)

	fmt.Println("\nTEST 4: Prime Sieve (10M)")
	benchSieve()

	fmt.Println("\nTEST 5: Vector Operations (10M)")
	benchVector(10000000)

	fmt.Println("\n=== BENCHMARK COMPLETE ===")
}

File: benchmarks/heavy_benchmark.cpp
// C++ HEAVY BENCHMARK
#include <iostream>
#include <chrono>
#include <vector>
#include <cstring>

using namespace std;
using namespace chrono;

// TEST 1: FIBONACCI
long long fib(int n) {
    if (n <= 1) return n;
    return fib(n - 1) + fib(n - 2);
}

long bench_fib() {
    auto start = high_resolution_clock::now();
    long long result = fib(40);
    auto end = high_resolution_clock::now();
    auto elapsed = duration_cast<milliseconds>(end - start).count();
    cout << "Fibonacci(40) = " << result << " in " << elapsed << "ms" << endl;
    return elapsed;
}

// TEST 2: MEMORY ALLOCATION
long bench_alloc(int count) {
    auto start = high_resolution_clock::now();
    vector<vector<char>> vecs;
    vecs.reserve(count);
    for (int i = 0; i < count; i++) {
        vecs.push_back(vector<char>(1024, i % 256));
    }
    auto end = high_resolution_clock::now();
    auto elapsed = duration_cast<milliseconds>(end - start).count();
    cout << "Allocated " << count << " x 1KB blocks in " << elapsed << "ms" << endl;
    return elapsed;
}

// TEST 3: CONCURRENT SIMULATION
long process_request(int id) {
    long sum = 0;
    for (int i = 0; i < 10000; i++) sum += i;
    return sum;
}

long bench_concurrent(int requests) {
    auto start = high_resolution_clock::now();
    int completed = 0;
    for (int i = 0; i < requests; i++) {
        if (process_request(i) > 0) completed++;
    }
    auto end = high_resolution_clock::now();
    auto elapsed = duration_cast<milliseconds>(end - start).count();
    cout << "Processed " << completed << " requests in " << elapsed << "ms" << endl;
    return elapsed;
}

// TEST 4: PRIME SIEVE
int sieve(int limit) {
    vector<bool> is_prime(limit, true);
    is_prime[0] = is_prime[1] = false;
    for (int i = 2; i * i < limit; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j < limit; j += i) {
                is_prime[j] = false;
            }
        }
    }
    int count = 0;
    for (bool p : is_prime) if (p) count++;
    return count;
}

long bench_sieve() {
    auto start = high_resolution_clock::now();
    int primes = sieve(10000000);
    auto end = high_resolution_clock::now();
    auto elapsed = duration_cast<milliseconds>(end - start).count();
    cout << "Found " << primes << " primes under 10M in " << elapsed << "ms" << endl;
    return elapsed;
}

// TEST 5: VECTOR OPERATIONS
long bench_vector(int size) {
    auto start = high_resolution_clock::now();
    vector<long> a(size), b(size), c(size);
    for (int i = 0; i < size; i++) {
        a[i] = i;
        b[i] = i * 2;
    }
    for (int i = 0; i < size; i++) {
        c[i] = a[i] + b[i];
    }
    auto end = high_resolution_clock::now();
    auto elapsed = duration_cast<milliseconds>(end - start).count();
    cout << "Vector add of " << size << " elements in " << elapsed << "ms" << endl;
    return elapsed;
}

int main() {
    cout << "=== C++ HEAVY BENCHMARK ===" << endl << endl;
    
    long total = 0;
    cout << "TEST 1: Fibonacci (Recursive)" << endl;
    total += bench_fib();
    
    cout << "\nTEST 2: Memory Allocation (1M blocks)" << endl;
    total += bench_alloc(1000000);
    
    cout << "\nTEST 3: Concurrent Requests (100K)" << endl;
    total += bench_concurrent(100000);
    
    cout << "\nTEST 4: Prime Sieve (10M)" << endl;
    total += bench_sieve();
    
    cout << "\nTEST 5: Vector Operations (10M)" << endl;
    total += bench_vector(10000000);
    
    cout << "\n=== TOTAL: " << total << "ms ===" << endl;
    return 0;
}
